Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrMetadataEmitter.kt

```diff
@@ -20,313 +21,460 @@ class StubIrMetadataEmitter(
-            underlyingType = element.aliasee.map(shouldExpandTypeAliases = false)
-            expandedType = element.aliasee.map()
-        }.let(typeAliases::add)
+    private fun writePackage(elements: KmElements) = KmPackage().also { km ->
+        km.fqName = packageFqName
+        km.typeAliases += elements.typeAliases.toList()
+        km.properties += elements.properties.toList()
+        km.functions += elements.functions.toList()
     }
 
-    override fun visitFunction(element: FunctionStub, data: StubContainer?) {
-        KmFunction(element.flags, element.name).apply {
-            element.annotations.mapTo(annotations, AnnotationStub::map)
-            returnType = element.returnType.map()
-            element.parameters.mapTo(valueParameters, FunctionParameterStub::map)
-            element.typeParameters.mapTo(typeParameters, TypeParameterStub::map)
-            uniqId = uniqIds.uniqIdForFunction(element)
-        }.let(functions::add)
+    /**
+     * StubIr translation result. Since Km* classes don't have common hierarchy we need
+     * to use list of Any.
+     */
+    private class KmElements(result: List<Any>) {
+        val classes: List<KmClass> = result.filterIsInstance<List<KmClass>>().flatten()
+        val properties: List<KmProperty> = result.filterIsInstance<KmProperty>()
+        val typeAliases: List<KmTypeAlias> = result.filterIsInstance<KmTypeAlias>()
+        val functions: List<KmFunction> = result.filterIsInstance<KmFunction>()
+        val constructors: List<KmConstructor> = result.filterIsInstance<KmConstructor>()
     }
 
-    override fun visitProperty(element: PropertyStub, data: StubContainer?) {
-        KmProperty(element.flags, element.name, element.getterFlags, element.setterFlags).apply {
-            element.annotations.mapTo(annotations, AnnotationStub::map)
-            uniqId = uniqIds.uniqIdForProperty(element)
-            returnType = element.type.map()
-            if (element.kind is PropertyStub.Kind.Var) {
-                val setter = element.kind.setter
-                setter.annotations.mapTo(setterAnnotations, AnnotationStub::map)
-                // TODO: Maybe it's better to explicitly add setter parameter in stub.
-                setterParameter = FunctionParameterStub("value", element.type).map()
+    /**
+     * Used to pass data between parents and children when visiting StubIr elements.
+     */
+    private data class VisitingContext(
+            val container: StubContainer? = null,
+            val uniqIds: StubIrUniqIdProvider,
+            val typeParametersInterner: Interner<TypeParameterStub> = Interner()
+    )
+
+    private val visitor = object : StubIrVisitor<VisitingContext, Any> {
+
+        override fun visitClass(element: ClassStub, data: VisitingContext): List<KmClass> {
+            val classVisitingContext = VisitingContext(
+                    container = element,
+                    uniqIds = data.uniqIds.createChild(element.classifier.relativeFqName),
```

Comment:
Is the entire `relativeFqName` required here?
Shouldn't it be just a last segment in such a name?